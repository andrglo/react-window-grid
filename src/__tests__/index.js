import React from 'react'
import {
  render,
  cleanup,
  // eslint-disable-next-line no-unused-vars
  prettyDOM
} from '@testing-library/react'

import {ReactWindowGrid} from '../..'

afterEach(cleanup)

const getStyle = container => container.getAttribute('style')

const tests = [
  [
    'Auto calculate height and width',
    {
      id: 'test1',
      height: 50,
      width: 100,
      columns: [
        {id: 'column1', label: 'Column 1'},
        {id: 'column2', label: 'Column 2'}
      ],
      columnHorizontalPadding: 8,
      recordset: [
        {column1: 'cell l/c 1/1', column2: 'cell 1/2'},
        {column1: 'cell 2/1', column2: 'cell 2/2'},
        {column1: 'cell 3/1', column2: 'cell 3/2'}
      ],
      rowHeaderWidth: 10
    }
  ],
  [
    'Case when recordset is empty use label or id to calc width',
    {
      id: 'test2',
      height: 50,
      width: 100,
      columns: [
        {id: 'column1', label: 'Column 1'},
        {id: 'column2', label: 'Column 2'}
      ],
      columnHorizontalPadding: 8,
      recordset: [],
      rowHeaderWidth: 10
    }
  ],
  [
    'Use customized render',
    {
      id: 'test3',
      height: 500,
      width: 200,
      columns: [
        {id: 'column1', label: 'Column 1'},
        {id: 'column2', label: 'Column 2'},
        {id: 'column3', label: 'Column 3'},
        {id: 'column4', label: 'Column 4'},
        {id: 'column5', label: 'Column 5'},
        {id: 'column6', label: 'Column 6'}
      ],
      columnHorizontalPadding: 8,
      recordset: [
        {
          column1: 'cell 1/1',
          column2: 'cell 1/2',
          column3: 'cell 1/3',
          column4: 'cell 1/4',
          column5: 'cell 1/5',
          column6: 'cell 1/6'
        },
        {
          column1: 'cell 2/1',
          column2: 'cell 2/2',
          column3: 'cell 2/3',
          column4: 'cell 2/4',
          column5: 'cell 2/5',
          column6: 'cell 2/6'
        },
        {
          column1: 'cell 3/1',
          column2: 'cell 3/2',
          column3: 'cell 3/3',
          column4: 'cell 3/4',
          column5: 'cell 3/5',
          column6: 'cell 3/6'
        },
        {
          column1: 'cell 4/1',
          column2: 'cell 4/2',
          column3: 'cell 4/3',
          column4: 'cell 4/4',
          column5: 'cell 4/5',
          column6: 'cell 4/6'
        },
        {
          column1: 'cell 5/1',
          column2: 'cell 5/2',
          column3: 'cell 5/3',
          column4: 'cell 5/4',
          column5: 'cell 5/5',
          column6: 'cell 5/6'
        },
        {
          column1: 'cell 6/1',
          column2: 'cell 6/2',
          column3: 'cell 6/3',
          column4: 'cell 6/4',
          column5: 'cell 6/5',
          column6: 'cell 6/6'
        }
      ],
      rowHeaderWidth: 30,
      columnHeaderRenderer: ({columnIndex, style}) => {
        const column = tests[2][1].columns[columnIndex]
        return <div style={style}>{column.label}</div>
      },
      rowHeaderRenderer: ({rowIndex, style}) => {
        return <div style={style}>{rowIndex + 1}</div>
      },
      cellRenderer: ({rowIndex, columnIndex, style}) => {
        const props = tests[2][1]
        const column = props.columns[columnIndex]
        const record = props.recordset[rowIndex]
        return <div style={style}>{record[column.id]}</div>
      }
    }
  ]
]

test(tests[0][0], async () => {
  const props = tests[0][1]
  const {container, getAllByText, rerender} = render(
    <ReactWindowGrid {...props} />
  )
  // console.log(prettyDOM(container))
  expect(container).toMatchSnapshot()

  const headers = getAllByText(/^Column \d$/)
  expect(headers.length).toBe(2)
  expect(getStyle(headers[0])).toMatch(/width: 20px;/)
  expect(getStyle(headers[1])).toMatch(/width: 16px;/)

  const rowHeaders = getAllByText(/^\d$/)
  expect(rowHeaders.length).toBe(3)

  let cells = getAllByText(/^cell/)
  expect(cells.length).toBe(6)
  expect(getStyle(cells[0])).toMatch(/width: 20px;/)
  expect(getStyle(cells[1])).toMatch(/width: 16px;/)
  expect(getStyle(cells[2])).toMatch(/width: 20px;/)
  expect(getStyle(cells[3])).toMatch(/width: 16px;/)
  expect(getStyle(cells[4])).toMatch(/width: 20px;/)
  expect(getStyle(cells[5])).toMatch(/width: 16px;/)

  const {height, ...rest} = props
  rerender(<ReactWindowGrid {...rest} maxHeight={39} />)
  cells = getAllByText(/^cell/)
  expect(cells.length).toBe(6)
  expect(getStyle(cells[0])).toMatch(/width: 20px;/)
  expect(getStyle(cells[1])).toMatch(/width: 16px;/)
  expect(getStyle(cells[2])).toMatch(/width: 20px;/)
  expect(getStyle(cells[3])).toMatch(/width: 16px;/)
  expect(getStyle(cells[4])).toMatch(/width: 20px;/)
  expect(getStyle(cells[5])).toMatch(/width: 16px;/)
})

test(tests[1][0], async () => {
  const props = tests[1][1]
  const {container, getAllByText, queryByText, rerender} = render(
    <ReactWindowGrid {...props} />
  )
  // console.log(prettyDOM(container))
  expect(container).toMatchSnapshot()

  const headers = getAllByText(/^Column \d$/)
  expect(headers.length).toBe(2)
  expect(getStyle(headers[0])).toMatch(/width: 16px;/)
  expect(getStyle(headers[1])).toMatch(/width: 16px;/)

  const rowHeaders = queryByText(/^\d$/)
  expect(rowHeaders).toBe(null)

  let cells = queryByText(/^cell/)
  expect(cells).toBe(null)

  const {height, ...rest} = tests[0][1]
  rerender(<ReactWindowGrid {...rest} maxHeight={39} />)
  cells = getAllByText(/^cell/)
  expect(cells.length).toBe(6)
  expect(getStyle(cells[0])).toMatch(/width: 20px;/)
  expect(getStyle(cells[1])).toMatch(/width: 16px;/)
  expect(getStyle(cells[2])).toMatch(/width: 20px;/)
  expect(getStyle(cells[3])).toMatch(/width: 16px;/)
  expect(getStyle(cells[4])).toMatch(/width: 20px;/)
  expect(getStyle(cells[5])).toMatch(/width: 16px;/)
})

test(tests[2][0], async () => {
  const props = tests[2][1]
  const {container, getAllByText, rerender} = render(
    <ReactWindowGrid {...props} />
  )
  // console.log(prettyDOM(container))
  expect(container).toMatchSnapshot()

  const headers = getAllByText(/^Column \d$/)
  expect(headers.length).toBe(6)
  expect(getStyle(headers[0])).toMatch(/width: 16px;/)
  expect(getStyle(headers[5])).toMatch(/width: 16px;/)

  const rowHeaders = getAllByText(/^\d$/)
  expect(rowHeaders.length).toBe(6)

  let cells = getAllByText(/^cell/)
  expect(cells.length).toBe(36)

  const {height, ...rest} = tests[0][1]
  rerender(<ReactWindowGrid {...rest} maxHeight={39} />)
  cells = getAllByText(/^cell/)
  expect(cells.length).toBe(6)
  expect(getStyle(cells[0])).toMatch(/width: 20px;/)
  expect(getStyle(cells[1])).toMatch(/width: 16px;/)
  expect(getStyle(cells[2])).toMatch(/width: 20px;/)
  expect(getStyle(cells[3])).toMatch(/width: 16px;/)
  expect(getStyle(cells[4])).toMatch(/width: 20px;/)
  expect(getStyle(cells[5])).toMatch(/width: 16px;/)
})
