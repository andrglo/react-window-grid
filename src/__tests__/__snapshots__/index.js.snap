// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auto calculate height and width 1`] = `
<div>
  <div
    id="test1"
    style="width: 100px; position: relative; height: 50px;"
  >
    <div
      style="position: absolute; top: 0px; left: 0px;"
    />
    <div
      style="position: absolute; top: 0px; left: 10px;"
    >
      <div
        style="position: relative; height: 12.5px; width: 90px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 100%; width: 36px;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 100%; width: 20px;"
          >
            Column 1
          </div>
          <div
            style="position: absolute; left: 20px; top: 0px; height: 100%; width: 16px;"
          >
            Column 2
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; left: 0px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 37.5px; width: 10px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 37.5px; width: 100%;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 12.5px; width: 100%;"
          >
            1
          </div>
          <div
            style="position: absolute; left: 0px; top: 12.5px; height: 12.5px; width: 100%;"
          >
            2
          </div>
          <div
            style="position: absolute; left: 0px; top: 25px; height: 12.5px; width: 100%;"
          >
            3
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; left: 10px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 37.5px; width: 90px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 37.5px; width: 36px;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 12.5px; width: 20px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell l/c 1/1
          </div>
          <div
            style="position: absolute; left: 20px; top: 0px; height: 12.5px; width: 16px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell 1/2
          </div>
          <div
            style="position: absolute; left: 0px; top: 12.5px; height: 12.5px; width: 20px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell 2/1
          </div>
          <div
            style="position: absolute; left: 20px; top: 12.5px; height: 12.5px; width: 16px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell 2/2
          </div>
          <div
            style="position: absolute; left: 0px; top: 25px; height: 12.5px; width: 20px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell 3/1
          </div>
          <div
            style="position: absolute; left: 20px; top: 25px; height: 12.5px; width: 16px; overflow: hidden; text-overflow: ellipsis;"
          >
            cell 3/2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Case when recordset is empty use label or id to calc width 1`] = `
<div>
  <div
    id="test2"
    style="width: 100px; position: relative; height: 50px;"
  >
    <div
      style="position: absolute; top: 0px; left: 0px;"
    />
    <div
      style="position: absolute; top: 0px; left: 10px;"
    >
      <div
        style="position: relative; height: 12.5px; width: 90px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 100%; width: 32px;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 100%; width: 16px;"
          >
            Column 1
          </div>
          <div
            style="position: absolute; left: 16px; top: 0px; height: 100%; width: 16px;"
          >
            Column 2
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; left: 0px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 37.5px; width: 10px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 0px; width: 100%;"
        />
      </div>
    </div>
    <div
      style="position: absolute; left: 10px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 37.5px; width: 90px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 0px; width: 100px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Use customized render 1`] = `
<div>
  <div
    id="test3"
    style="width: 200px; position: relative; height: 500px;"
  >
    <div
      style="position: absolute; top: 0px; left: 0px;"
    />
    <div
      style="position: absolute; top: 0px; left: 30px;"
    >
      <div
        style="position: relative; height: 12.5px; width: 170px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 100%; width: 96px;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 100%; width: 16px;"
          >
            Column 1
          </div>
          <div
            style="position: absolute; left: 16px; top: 0px; height: 100%; width: 16px;"
          >
            Column 2
          </div>
          <div
            style="position: absolute; left: 32px; top: 0px; height: 100%; width: 16px;"
          >
            Column 3
          </div>
          <div
            style="position: absolute; left: 48px; top: 0px; height: 100%; width: 16px;"
          >
            Column 4
          </div>
          <div
            style="position: absolute; left: 64px; top: 0px; height: 100%; width: 16px;"
          >
            Column 5
          </div>
          <div
            style="position: absolute; left: 80px; top: 0px; height: 100%; width: 16px;"
          >
            Column 6
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; left: 0px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 487.5px; width: 30px; overflow: hidden; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 75px; width: 100%;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 12.5px; width: 100%;"
          >
            1
          </div>
          <div
            style="position: absolute; left: 0px; top: 12.5px; height: 12.5px; width: 100%;"
          >
            2
          </div>
          <div
            style="position: absolute; left: 0px; top: 25px; height: 12.5px; width: 100%;"
          >
            3
          </div>
          <div
            style="position: absolute; left: 0px; top: 37.5px; height: 12.5px; width: 100%;"
          >
            4
          </div>
          <div
            style="position: absolute; left: 0px; top: 50px; height: 12.5px; width: 100%;"
          >
            5
          </div>
          <div
            style="position: absolute; left: 0px; top: 62.5px; height: 12.5px; width: 100%;"
          >
            6
          </div>
        </div>
      </div>
    </div>
    <div
      style="position: absolute; left: 30px; top: 12.5px;"
    >
      <div
        style="position: relative; height: 487.5px; width: 170px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 75px; width: 96px;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 0px; height: 12.5px; width: 16px;"
          >
            cell 1/6
          </div>
          <div
            style="position: absolute; left: 0px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 12.5px; height: 12.5px; width: 16px;"
          >
            cell 2/6
          </div>
          <div
            style="position: absolute; left: 0px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 25px; height: 12.5px; width: 16px;"
          >
            cell 3/6
          </div>
          <div
            style="position: absolute; left: 0px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 37.5px; height: 12.5px; width: 16px;"
          >
            cell 4/6
          </div>
          <div
            style="position: absolute; left: 0px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 50px; height: 12.5px; width: 16px;"
          >
            cell 5/6
          </div>
          <div
            style="position: absolute; left: 0px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/1
          </div>
          <div
            style="position: absolute; left: 16px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/2
          </div>
          <div
            style="position: absolute; left: 32px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/3
          </div>
          <div
            style="position: absolute; left: 48px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/4
          </div>
          <div
            style="position: absolute; left: 64px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/5
          </div>
          <div
            style="position: absolute; left: 80px; top: 62.5px; height: 12.5px; width: 16px;"
          >
            cell 6/6
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
